crypto/provider_core.c

```diff
@@ -1401,8 +1401,10 @@ int ossl_provider_doall_activated(OSSL_LIB_CTX *ctx,
     for (curr = 0; curr < max; curr++) {
         OSSL_PROVIDER *prov = sk_OSSL_PROVIDER_value(provs, curr);
 
-        if (!cb(prov, cbdata))
+        if (!cb(prov, cbdata)) {
+            curr = -1;
```

Comment:
It seems strange to iterate thru all of them... This assumes the cb will fail when curr is zero. If it somehow passed on curr=0 then failed on cur=1 then this is not right. cur--; would be more correct in this case.. since that willl increment it back to the curr value that failed.

Reply:
No, that's not correct. It needs to iterate through all of them. If we've got to this point we've already successfully iterated through the whole set upping the ref count in the loop starting on line 1362. If we get to this point we are now iterating again calling the cb on each entry. If at any point we fail, we need to go back to the start and iterate through them all again releasing the ref we obtained earlier.

Reply:
In an earlier iteration of this code, `sk_OSSL_PROVIDER_pop` was used, not `sk_OSSL_PROVIDER_value`. With the latter, `curr` has no inherent value per se here, which becomes clear when you realize that when coming of this loop, it will always have the value -1.

That should also make it clear that the loop at `finish:` can be refactored into a simple `sk_OSSL_PROVIDER_pop_free` call at this point.

Reply:
> That should make it clear that the loop at `finish:` can be refactored into a simple `sk_OSSL_PROVIDER_pop_free` call at this point.

Not in the case you go to the same code by the `goto err_unlock;` in the above loop.

Reply:
> That should make it clear that the loop at finish: can be refactored into a simple sk_OSSL_PROVIDER_pop_free call at this point.

As @t8m points out, we cannot use `sk_OSSL_PROVIDER_pop_free` because of the goto on line 1388